src/api-service/__app__/onefuzzlib/azure/nsg.py

```
@@ -0,0 +1,484 @@
+    try:
+        nsg = network_client.network_security_groups.get(resource_group, name)
+        return cast(NetworkSecurityGroup, nsg)
+    except (ResourceNotFoundError, CloudError) as err:
+        logging.debug("nsg %s does not exist: %s", name, err)
+        return None
+
+
+def create_nsg(name: str, location: Region) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+
+    logging.info("creating nsg %s:%s:%s", resource_group, location, name)
+    network_client = get_network_client()
+
+    params: Dict = {
+        "location": location,
+    }
+
+    if "ONEFUZZ_OWNER" in os.environ:
+        params["tags"] = {"OWNER": os.environ["ONEFUZZ_OWNER"]}
+
+    try:
+        network_client.network_security_groups.begin_create_or_update(
+            resource_group, name, params
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if is_concurrent_request_error(str(err)):
+            logging.debug(
+                "create NSG had conflicts with concurrent request, ignoring %s", err
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_CREATE,
+            errors=["Unable to create nsg %s due to %s" % (name, err)],
+        )
+    return None
+
+
+def list_nsgs() -> List[NetworkSecurityGroup]:
+    resource_group = get_base_resource_group()
+    network_client = get_network_client()
+    return list(network_client.network_security_groups.list(resource_group))
+
+
+def update_nsg(nsg: NetworkSecurityGroup) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+
+    logging.info("updating nsg %s:%s:%s", resource_group, nsg.location, nsg.name)
+    network_client = get_network_client()
+
+    try:
+        network_client.network_security_groups.begin_create_or_update(
+            resource_group, nsg.name, nsg
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if is_concurrent_request_error(str(err)):
+            logging.debug(
```

Again, is this the right log level? How should we log when catching these kinds of errors?

---

This error is ignored, since there is a request in progress already. In this specific case "debug" makes sense

---

Existing components log with debug for concurrent requests:
https://github.com/microsoft/onefuzz/blob/c82c84d1a4fa6f5c9faf57a605fa83afa2f7fbf6/src/api-service/__app__/onefuzzlib/azure/vmss.py#L376

https://github.com/microsoft/onefuzz/blob/c82c84d1a4fa6f5c9faf57a605fa83afa2f7fbf6/src/api-service/__app__/onefuzzlib/azure/vm.py#L117

---

That makes sense.